File: src/main/scala/info/crlog/higgs/protocols/boson/json/BosonServer.scala
package info.crlog.higgs.protocols.boson.json

import info.crlog.higgs.RPCServer
import io.netty.channel.{Channel, ChannelHandlerContext}
import java.io.Serializable
import info.crlog.higgs.util.StringUtil
import org.codehaus.jackson.map.JsonMappingException

/**
 * @author Courtney Robinson <courtney@crlog.info>
 */
class BosonServer(host: String, port: Int, compress: Boolean = false)
  extends RPCServer[Message](host, port, compress) {

  val serializer = new BosonSerializer()

  def decoder() = new BosonDecoder()

  def encoder() = new BosonEncoder()

  def allTopicsKey(): String = ""

  def message(context: ChannelHandlerContext, value: Array[Byte]) {
    try {
      val data = serializer.deserialize(value)
      val size = notifySubscribers(context.channel(),
        data.topic,
        data
      )
      if (size == 0) {
        respond(context.channel(),
          new Message(data.callback.getOrElse("error"),None,None,Map("msg" -> "Method %s not found".format(data.topic), "error" -> "not_found"))
        )
      }
    } catch {
      case e: JsonMappingException => {
        val data = StringUtil.getString(value)
        log.warn("Unable to deserialize message %s".format(data), e)
        respond(context.channel(),
          new Message("json_deserialize_error", None,None,
            Map("msg" -> "Incomplete or Invalid JSON received %s ".format(data))
          )
        )
      }
    }
  }

  def broadcast(obj: Message) {}

  def getArguments(param: Message): Seq[AnyRef] = {
    param.arguments match {
      case None => Seq()
      case Some(arr) => arr
    }
  }

  def clientCallback(param: Message): String = {
    param.callback match {
      case None => ""
      case Some(c) => c
    }
  }

  def newResponse(remoteMethodName: String, clientCallbackID: String,
                  response: Option[Serializable], error: Option[Throwable]) = {
//    val data: Map[String, Any] = Map(
//      "remote_method" -> remoteMethodName
//    ) ++
//      (response match {
//        case None => Map()
//        case Some(r) => Map("response" -> r)
//      }) ++
//      (error match {
//        case None => Map()
//        case Some(e) => Map("error" -> e)
//      })

    new Message(clientCallbackID,None,None, response.getOrElse(null))
  }

  def listen[T, U](topic: String, callback: (T) => U)(implicit mf: Manifest[T]) {
    listen(topic, (msg: T, i: InvalidBosonResponse) => {
      if (i != null) {
        log.warn("Invalid boson message received ", i)
      } else {
        callback(msg)
      }
    })

  }

  /**
   * Listen for a topic being broadcasted
   * @param topic
   * @param callback
   * @param mf
   * @tparam T
   * @tparam U
   * @return
   */
  def listen[T, U](topic: String, callback: (T, InvalidBosonResponse) => U)
                  (implicit mf: Manifest[T]) {
    listen(topic, (c: Channel, m: Message) => {
      serializer.unmarshal(m, (msg: T, i: InvalidBosonResponse) => {
        callback(msg, i)
      })
    })
  }
}

File: higgs-boson/src/main/java/io/higgs/boson/BosonServer.java
package io.higgs.boson;

import com.google.common.base.Optional;
import io.higgs.RPCServer;
import io.higgs.boson.serialization.BosonDecoder;
import io.higgs.boson.serialization.BosonEncoder;
import io.higgs.boson.serialization.mutators.ReadMutator;
import io.higgs.boson.serialization.mutators.ReadWriteMutator;
import io.higgs.boson.serialization.mutators.WriteMutator;
import io.higgs.boson.serialization.v1.BosonReader;
import io.higgs.boson.serialization.v1.BosonWriter;
import io.higgs.events.ChannelMessage;
import io.netty.buffer.ByteBuf;
import io.netty.channel.Channel;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelPipeline;

import java.util.HashSet;
import java.util.Set;

/**
 * @author Courtney Robinson <courtney@crlog.info>
 */
public class BosonServer extends RPCServer<BosonMessage, BosonMessage, ByteBuf> {
    protected boolean compression;
    protected boolean ssl;
    protected Set<WriteMutator> writeMutators = new HashSet<>();
    protected Set<ReadMutator> readMutators = new HashSet<>();

    public BosonServer(int port) {
        this(port, false);
        setEnableProtocolSniffing(false);
    }

    public BosonServer(int port, boolean compression) {
        this(port, compression, false);
    }

    public BosonServer(int port, boolean compression, boolean ssl) {
        super(port);
        this.compression = compression;
        this.ssl = ssl;
    }

    public Object[] getArguments(final Class<?>[] argTypes, ChannelMessage<BosonMessage> request) {
        return request.message.arguments;
    }

    protected BosonMessage newResponse(String methodName, ChannelMessage<BosonMessage> request,
                                       Optional<Object> returns, Optional<Throwable> error) {
        BosonMessage msg = new BosonMessage();
        msg.method = request.message.callback;
        Object[] args = new Object[2];
        args[0] = null;
        if (returns.isPresent()) {
            args[0] = returns.get();
        }
        args[1] = null;
        if (error.isPresent()) {
            args[1] = error.get();
        }
        msg.arguments = args;
        return msg;
    }

    @Override
    public ByteBuf serialize(final Channel ctx, final BosonMessage msg) {
        return new BosonWriter(readMutators, msg).serialize();
    }

    @Override
    public BosonMessage deserialize(final ChannelHandlerContext ctx, final ByteBuf msg) {
        return new BosonReader(writeMutators, msg).deSerialize();
    }

    @Override
    public String getTopic(final BosonMessage msg) {
        return msg.method;
    }

    @Override
    protected boolean setupPipeline(final ChannelPipeline pipeline) {
        pipeline.addLast("decoder", new BosonDecoder());
        pipeline.addLast("encoder", new BosonEncoder());
        return true; //auto add handler
    }

    public boolean addWriteMutator(WriteMutator mutator) {
        return mutator != null && writeMutators.add(mutator);
    }

    public boolean addReadMutator(ReadMutator mutator) {
        return mutator != null && readMutators.add(mutator);
    }

    /**
     * Add a mutator which can be used by both the serializer and de-serializer
     *
     * @param mutator the mutator to add
     * @return true if not null and  successfully added
     */
    public boolean addReadWriteMutator(ReadWriteMutator mutator) {
        return mutator != null && readMutators.add(mutator) && writeMutators.add(mutator);
    }
}
